blog/2021/01/snoopi_deep.md

```diff
@@ -0,0 +1,221 @@
+
+The `ROOT` node is a bit different: its exclusive time measures the time spent on all operations *except* inference.
+In this case, we see that the entire call took approximately 10ms, of which 9.3ms was spent on activities besides inference.
+Almost all of that was code-generation, but it also includes the time needed to run the code.
+Just 0.76ms was needed to run type-inference on this entire series of calls.
+As you will quickly discover as you use `@snoopi_deep`, inference takes much more time on more complicated code.
+
+You can extract the `MethodInstance` with
+
+```
+julia> Core.MethodInstance(tinf)
+MethodInstance for ROOT()
+
+julia> Core.MethodInstance(tinf.children[1])
+MethodInstance for packintype(::Int64)
+```
+
+## Visualizing the output
+
+We can also display this tree as a flame graph, using the [ProfileView](https://github.com/timholy/ProfileView.jl) package:
+
+```
+julia> fg = flamegraph(tinf)
+Node(FlameGraphs.NodeData(ROOT() at typeinfer.jl:75, 0x00, 0:10080857))
+
+julia> using ProfileView
+
+julia> ProfileView.view(fg)
+```
+
+You should see something like this:
+
+![flamegraph](/assets/blog/2021-latency/flamegraph-flatten-demo.png)
+
+Users are encouraged to read the ProfileView documentation to understand how to interpret this, but briefly:
+
+- the horizontal axis is time (wide boxes take longer than narrow ones), the vertical axis is call depth
+- hovering over a box displays the method that was inferred
+- left-clicking on a box causes the full MethodInstance to be printed in your REPL session
+- right-clicking on a box opens the corresponding method in your editor (you must have `ENV["EDITOR"]` configured appropriately for this to work)
+- ctrl-click can be used to zoom in (you can do rubber band selection or zoom in/zoom out)
+- empty horizontal spaces correspond to activities other than type-inference; in this case, the relatively narrow flame followed by a lot of empty space indicates that all the type inference happened at the beginning and accounted for only a small fraction of the total time.
+
+You can explore this flamegraph and compare it to the output from `print_tree`.
+
+In less trivial examples, these flame graphs may look more interesting:
+
+![flamegraph-complex](/assets/blog/2021-latency/flamegraph-complex.png)
+
+Here, the red boxes correspond to `MethodInstance`s which cannot be "naturally" precompiled.
+This occurs when the method is owned by one module but the argument types are from another unrelated module.
+We'll see how to deal with these in later installments.
+
+You also see breaks in the flamegraph. During these periods, code generation and runtime create new objects and then call methods on those objects; if one of those calls requires a fresh entrance into inference, that triggers the creation of a new flame.
+Hence, the number of distinct flames (which is just equal to `length(tinf.children)`) gives you a rough indication of how frequently the chains of inference were broken.
```

Comment:
```suggestion
Hence, the number of distinct flames (which is just equal to `length(tinf.children)`) gives you a rough indication of how frequently the chains of inference were broken. This can be caused by type instability, separate top-level calls from the REPL during `@snoopi_deep`, calls to `eval` in your code, or some other cause for julia to start running type inference.
```